### PR TITLE
Fix param expansion in inherited environment.

### LINKED ISSUE #4242: parameters: inheriting from unrelated families
Inheritance works with task parameters e.g:

```ini
[runtime]
    [[FAM<param>]]
    [[task<param>]]
        inherit = FAM<param>
```

However, when mixing parameters across inheritance it only partially works.

```ini
[runtime]
    [[FAM<param1>]]
    [[task<param2>]]
        inherit = FAM<param1>
```

### Example

```ini
[task parameters]
    a = 1..2
    b = 1..2

[scheduling]
    [[graph]]
        R1 = """
            <a>
        """

[runtime]
    [[<a>]]
        inherit = <b=1>
        script = test $b -eq 1

    [[<b>]]
        [[[environment]]]
            b = %(b)d
```

The inheritance does indeed work as shown by the output of `cylc config`:

```console
$ cylc config --sparse  param-test 
[task parameters]
    a = 1, 2
    b = 1, 2
[scheduling]
    cycling mode = integer
    initial cycle point = 1
    final cycle point = 1
    [[graph]]
        R1 = <a>
[runtime]
    [[root]]
    [[_a1]]
        inherit = _b1
        script = test $b -eq 1
        [[[environment]]]
            b = %(b)d
    [[_a2]]
        inherit = _b1
        script = test $b -eq 1
        [[[environment]]]
            b = %(b)d
    [[_b1]]
        [[[environment]]]
            b = %(b)d
    [[_b2]]
        [[[environment]]]
            b = %(b)d
[visualization]
    [[node attributes]]
```

However, the parameter in the environment variable is not expanded in the job script:

```bash
# job

cylc__job__inst__user_env() {
    # TASK RUNTIME ENVIRONMENT:
    export b
    b="%(b)d"
}
```

### Use Case

Example use case which involves mapping one parameter onto another:

```ini
[runtime]
    # define environments
    [[environment<environment>]]

    # map models onto environments
    [[model<model=a>]]
        inherit = environment<1>
    [[model<model=b>]]
        inherit = environment<2>
    [[model<model=c>]]
        inherit = environment<1>
```

### PR DESCRIPTION
**Pull request by hjoliver**
These changes close #4242 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] No dependency changes. 
- [x] Appropriate tests are included (extended an existing functional test).
- [x] Appropriate change log entry included.
- [x] No documentation update required (this branch conforms to advertised behaviour).

### CODE REVIEW

**Comment by hjoliver** [comment]
This is ready to go now.

**Review by kinow — APPROVED** [review]
Code and tests look good! Pending only CI to pass :+1:

**Comment by kinow** [comment]
mypy error fixed in the merged #4251

**Comment by hjoliver** [comment]
Rebased.

**Comment by kinow** [comment]
@hjoliver can we ignore the macos failure?

```
2021-06-09T01:59:36.3004720Z TOTAL                                                20623   7211   8643      0    75.36%
2021-06-09T01:59:36.3005000Z 
2021-06-09T01:59:36.3005400Z =========================== short test summary info ============================
2021-06-09T01:59:36.3006860Z FAILED tests/integration/test_scan.py::test_scan_sigstop - AssertionError: as...
2021-06-09T01:59:36.3007580Z ================== 1 failed, 86 passed, 24 warnings in 33.58s ==================
```

```
2021-06-09T01:59:36.2794800Z E           AssertionError: assert [(20, '[one.1...05056910e93')] == [(30, 'Workfl...05056910e93')]
2021-06-09T01:59:36.2796570Z E             At index 0 diff: (20, '[one.1] -released from runahead') != (30, 'Workflow not running: 46dcaa6a-c8c6-11eb-8e68-005056910e93')
2021-06-09T01:59:36.2798370Z E             Left contains one more item: (30, 'Workflow not running: 46dcaa6a-c8c6-11eb-8e68-005056910e93')
2021-06-09T01:59:36.2799580Z E             Full diff:
2021-06-09T01:59:36.2800190Z E               [
2021-06-09T01:59:36.2800750Z E             +  (20,
2021-06-09T01:59:36.2801840Z E             +   '[one.1] -released from runahead'),
2021-06-09T01:59:36.2802600Z E                (30,
2021-06-09T01:59:36.2803850Z E                 'Workflow not running: 46dcaa6a-c8c6-11eb-8e68-005056910e93'),
2021-06-09T01:59:36.2804740Z E               ]
```

**Comment by hjoliver** [comment]
Yeah that MacOS scan test is failing on most PRs now, nothing to do with this change.  Ping @oliver-sanders

**Review by kinow — APPROVED** [review]

File: cylc/flow/config.py [hunk 1]
```diff
@@ -991,10 +991,11 @@ def _expand_runtime(self):
...
                         if used_indices:
-                            if name not in self.task_param_vars:
-                                self.task_param_vars[name] = used_indices
-                            else:
-                                self.task_param_vars[name].update(used_indices)
+                            self.task_param_vars.setdefault(
+                                name, {}
+                            ).update(
+                                used_indices
+                            )
```
Comment (kinow) [inline cylc/flow/config.py:998]: :+1:

**Review by kinow — APPROVED** [review]

**Review by oliver-sanders — CHANGES_REQUESTED** [review]
Sadly this didn't work for my example workflow, took a bit of staring before I found the edge case that was causing the bug.

On this branch the following example fails:

```ini
[task parameters]
    a = 1..3
    b = 1

[scheduling]
    [[graph]]
        R1 = x<b>

[runtime]
    [[FAM<a>]]
        [[[environment]]]
            param = %(a)s

    [[x<b=1>]]
        inherit = FAM<a=1>
        script = test $param -eq 1

    [[x<b>]]
        pre-script = echo pre
```

However, if you comment out the last task:

```
    # [[x<b>]]
    #    pre-script = echo pre
```

It will pass on this branch (but not on master).

**Comment by hjoliver** [comment]
God damn it. Sadly indeed!

**Comment by hjoliver** [comment]
It certainly works for your original example, on #4242

**Comment by hjoliver** [comment]
Interestingly, it works if I swap the order:
```
    [[x<b>]]
        pre-script = echo pre

    [[x<b=1>]]
        inherit = FAM<a=1>
        script = test $param -eq 1
```

**Comment by hjoliver** [comment]
Fixed now, and test extended to cover this case.

**Review by oliver-sanders — APPROVED** [review]
My example now passes, thanks.

**Pull request merged**